include/exadg/postprocessor/kinetic_energy_spectrum.cpp

```diff
@@ -403,100 +408,54 @@ KineticEnergySpectrumCalculator<dim, Number>::setup(
 }
 
 template<int dim, typename Number>
-void
-KineticEnergySpectrumCalculator<dim, Number>::evaluate(VectorType const & velocity,
-                                                       double const &     time,
-                                                       int const &        time_step_number)
+bool
+KineticEnergySpectrumCalculator<dim, Number>::needs_evaluation(double const       time,
+                                                               unsigned int const time_step_number)
 {
-  if(data.calculate == true)
-  {
-    if(time_step_number >= 0) // unsteady problem
-    {
-      if(needs_to_be_evaluated(time, time_step_number))
-      {
-        if(data.exploit_symmetry)
-        {
-          unsigned int n_cells_1d =
-            data.n_cells_1d_coarse_grid * dealii::Utilities::pow(2, data.refine_level);
-
-          velocity_full = std::make_shared<VectorType>();
-          initialize_dof_vector(*velocity_full, *dof_handler_full);
-
-          apply_taylor_green_symmetry(*dof_handler,
-                                      *dof_handler_full,
-                                      n_cells_1d,
-                                      data.length_symmetric_domain /
-                                        static_cast<double>(n_cells_1d),
-                                      velocity,
-                                      *velocity_full);
-
-          do_evaluate(*velocity_full, time);
-        }
-        else
-        {
-          do_evaluate(velocity, time);
-        }
-      }
-    }
-    else // steady problem (time_step_number = -1)
-    {
-      AssertThrow(
-        false,
-        dealii::ExcMessage(
-          "Calculation of kinetic energy spectrum only implemented for unsteady problems."));
-    }
-  }
+  return time_control.needs_evaluation(time, time_step_number);
 }
 
 template<int dim, typename Number>
-bool
-KineticEnergySpectrumCalculator<dim, Number>::needs_to_be_evaluated(
-  double const       time,
-  unsigned int const time_step_number)
+void
+KineticEnergySpectrumCalculator<dim, Number>::evaluate(VectorType const & velocity,
+                                                       double const       time,
+                                                       int const          time_step_number)
 {
-  bool evaluate = false;
+  AssertThrow(time_control.get_last_timestep_number_to_evaluate() == time_step_number,
+              dealii::ExcMessage(
+                "evaluate should only be called once needs_evaluation() returned true."));
```

Comment:
can we get rid of this assert? My hope would be that we no longer have to pass `time_step_number` to `evaluate()` functions. We can achieve this, apart from the assert, by adding a boolean parameter `evaluate(..., bool const steady_problem)` only for those postprocessing modules that are implemented also for steady problems. The function is then called with `evaluate(..., time_step_number < 0)` (or whatever the convention for steady problems is). If this is not implemented (as in this case here), the function `evaluate()` just contains the vector and the time.

Reply:
We can do this if we assume nobody is using this in external projects. Not changing the interface results in output at every timestep. If this is ok for you I will remove it.

Reply:
That's acceptable in my opinion. For external projects nothing is "falsified", just an if-statement has to be introduced. I think there are good reasons why we make the changes of this PR, i.e. it turned out that we had a bad design before.

Reply:
[d0bc416](https://github.com/exadg/exadg/pull/262/commits/d0bc416c46ebf80a9ec7966064f64303d03a538b)